pkg/storage/olympus/getter.go

```diff
@@ -1,80 +1,82 @@
-
-	// fetch mod file
-	var mod []byte
-
-	modResp, err := s.client.Get(modURI)
-	if err != nil {
-		return nil, err
-	}
-	defer modResp.Body.Close()
-	if modResp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("GET %v returned unexpected status: %v", modURI, modResp.StatusCode)
-	}
-
-	mod, err = ioutil.ReadAll(modResp.Body)
+	resp, err := s.client.Get(infoURI)
 	if err != nil {
-		return nil, err
+		return nil, errors.E(op, err)
 	}
-
-	// fetch source file
-	var zip []byte
-	zipResp, err := s.client.Get(zipURI)
+	defer resp.Body.Close()
+	info, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return nil, err
+		return nil, errors.E(op, err)
 	}
-	defer zipResp.Body.Close()
-	if zipResp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("GET %v returned unexpected status: %v", zipURI, zipResp.StatusCode)
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.E(op, fmt.Sprintf("unexpected status %v - body: %s", resp.StatusCode, info), resp.StatusCode)
 	}
 
-	zip, err = ioutil.ReadAll(zipResp.Body)
+	return info, nil
+}
+
+// GoMod storage.Getter implementation
+func (s *ModuleStore) GoMod(ctx context.Context, module, vsn string) ([]byte, error) {
+	const op errors.Op = "olympus.GoMod"
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "storage.olympus.GoMod")
+	defer sp.Finish()
+
+	// TODO: fetch from endpoint
+	modURI := fmt.Sprintf("%s/%s", s.url, config.PackageVersionedName(module, vsn, "mod"))
+	resp, err := s.client.Get(modURI)
 	if err != nil {
-		return nil, err
+		return nil, errors.E(op, err)
 	}
-
-	// fetch info file
-	var info []byte
-	infoResp, err := s.client.Get(infoURI)
+	defer resp.Body.Close()
+	mod, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return nil, err
+		return nil, errors.E(op, err)
 	}
-	defer infoResp.Body.Close()
-	if infoResp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("GET %v returned unexpected status: %v", infoURI, infoResp.StatusCode)
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.E(op, fmt.Sprintf("unexpected status %v - body: %s", resp.StatusCode, mod), resp.StatusCode)
 	}
 
-	info, err = ioutil.ReadAll(infoResp.Body)
+	return mod, err
+}
+
+// Zip storage.Getter implementation
+func (s *ModuleStore) Zip(ctx context.Context, module, vsn string) (io.ReadCloser, error) {
+	const op errors.Op = "olympus.Zip"
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "storage.olympus.Zip")
+	defer sp.Finish()
+
+	// TODO: fetch from endpoint
+	zipURI := fmt.Sprintf("%s/%s", s.url, config.PackageVersionedName(module, vsn, "zip"))
+	resp, err := s.client.Get(zipURI)
 	if err != nil {
 		return nil, err
 	}
+	if resp.StatusCode != http.StatusOK {
+		msg := fmt.Errorf("GET %v returned unexpected status: %v", zipURI, resp.StatusCode)
+		return nil, errors.E(op, msg)
+	}
 
-	return &storage.Version{
-		Info: info,
-		Mod:  mod,
-		Zip:  ioutil.NopCloser(bytes.NewReader(zip)),
-	}, nil
+	return resp.Body, nil
```

Comment:
i'm on the fence here. while this is readcloser and it is correct to return it, it will keep socket opened until zip is processed which might easily lead to `out of sockets`

Reply:
@michalpristas part of my clean up is that I'd like to suggest removing this whole package. issues 349 and 309 will make it so that we always redirect to Olympus so we never have to "stream from it". And even if we do, it will be in the shape of a download.Protocol interface, which goes to our Proxy storage first then back to the user. Let me know if that does or doesn't make sense :)

Reply:
Yes it does @marwan-at-work Thanks.